napari/utils/key_bindings.py

```diff
@@ -182,6 +187,257 @@ def normalize_key_combo(key_combo):
+    ]
+    ctx = {k: v for k, v in context.items() if k in context_keys}
+    return function(**ctx)
+
+
+@dataclass
+class Action:
+    command: Callable
+    description: str
+    keymappable: Any
+
+
+class ActionManager:
+    """
+    Manage the bindings between buttons; shortcuts, callbacks gui elements...
+
+    The action manager is aware of the various buttons, keybindings and other
+    elements that may trigger an action and is able to synchronise all of those.
+    Thus when a shortcut is bound; this should be capable of updating the
+    buttons tooltip menus etc to show the shortcuts, descriptions...
+
+    In most cases this should also allow to bind non existing shortcuts,
+    actions, buttons, in which case they will be bound only once the actions are
+    registered.
+
+    As the action manager also knows about all the UI elements that can trigger
+    an action, it can also be used to flash the buttons when the action is
+    triggered separately.
+
+    For actions that need access to a global element (a viewer, a plugin, ... ),
+    you want to give this item a unique name, and add it to the action manager
+    `context` object.
+
+    >>> action_manager.context['number'] = 1
+    ... action_manager.context['qtv'] = viewer.qt_viewer
+
+    >>> def callback(qtv, number):
+    ...     qtv.dims[number] +=1
+
+    >>> action_manager.register_action('bump one', callback,
+    ...     'Add one to dims',
+    ...     None)
+
+    The callback signature is going to be inspected and required globals passed
+    in.
+    """
+
+    _actions: Dict[str, Action]
+
+    def __init__(self):
+        # map associating a name/id with a Comm
+        self._actions = {}
+        self._buttons = defaultdict(lambda: [])
+        self._qactions = defaultdict(lambda: [])
+        self._shortcuts = {}
+        self._orphans = {}
+        self.context = {}
+
+    def register_action(self, name, command, description, keymappable):
+        """
+        Register an action for future usage
+
+        An action is generally a callback associated with
+         - a name (unique),
+         - a description
+         - A keymapable (easier for focus and backward compatibility).
+
+        Actions can then be later bound/unbound from button elements, and
+        shortcuts; and the action manager will take care of modifying the keymap
+        of instances to handle shortcuts; and UI elements to have tooltips with
+        descriptions and shortcuts;
+
+        Parameters
+        ----------
+
+        name: str
+            unique name/id of the command that can be used to refer to this
+            command
+        command: callable
+            take 0, or 1 parameter; is `instance` is not None, will be called
+            `instance` as first parameter.
+        description: str
+            Long string to describe what the command does, will be used in
+            tooltips.
+
+        """
+        assert name not in self._actions, name
```

Comment:
If assertions are disabled (for whatever reason) this would not work. Maybe use a good ol' if plus exception raising?